### PR TITLE
Change TestInstanceAutoJoin to adapt to cloud environment

### LINKED ISSUE #1260: Test TestInstanceAutoJoin.testAutoRegistrationShouldFailWhenWaitingResponse is flaky
- Flaky
LOG:
>020-08-12T12:38:16.1190369Z [ERROR] testAutoRegistrationShouldFailWhenWaitingResponse(org.apache.helix.integration.paticipant.TestInstanceAutoJoin)  Time elapsed: 0.439 s  <<< FAILURE!
2020-08-12T12:38:16.1194078Z java.lang.AssertionError: expected:<true> but was:<false>
2020-08-12T12:38:16.1202161Z 	at org.apache.helix.integration.paticipant.TestInstanceAutoJoin.testAutoRegistrationShouldFailWhenWaitingResponse(TestInstanceAutoJoin.java:113)

>-08-08T10:12:24.8433385Z [ERROR] testAutoRegistrationShouldFailWhenWaitingResponse(org.apache.helix.integration.paticipant.TestInstanceAutoJoin)  Time elapsed: 0.366 s  <<< FAILURE!
2020-08-08T10:12:24.8433664Z java.lang.AssertionError: expected:<true> but was:<false>
2020-08-08T10:12:24.8434622Z 	at org.apache.helix.integration.paticipant.TestInstanceAutoJoin.testAutoRegistrationShouldFailWhenWaitingResponse(TestInstanceAutoJoin.java:113)

Theory and proof:
Running in local environment, the code would always work. With exception like:
>248277 [ClusterManager_Watcher_CLUSTER_TestInstanceAutoJoin_localhost_279700_PARTICIPANT_10008] ERROR org.apache.helix.manager.zk.ZKHelixManager  - fail to createClient.
org.apache.helix.HelixException: Failed to get Azure cloud instance information from url {}
	at org.apache.helix.cloud.azure.AzureCloudInstanceInformationProcessor.getAzureCloudInformationFromUrl(AzureCloudInstanceInformationProcessor.java:122)
	at org.apache.helix.cloud.azure.AzureCloudInstanceInformationProcessor.fetchCloudInstanceInformation(AzureCloudInstanceInformationProcessor.java:95)
	at org.apache.helix.manager.zk.ParticipantManager.getCloudInstanceInformation(ParticipantManager.java:225)
	at org.apache.helix.manager.zk.ParticipantManager.joinCluster(ParticipantManager.java:201)
	at org.apache.helix.manager.zk.ParticipantManager.handleNewSession(ParticipantManager.java:141)
	at org.apache.helix.manager.zk.ZKHelixManager.handleNewSessionAsParticipant(ZKHelixManager.java:1359)
	at org.apache.helix.manager.zk.ZKHelixManager.handleNewSession(ZKHelixManager.java:1319)
	at org.apache.helix.manager.zk.ZKHelixManager.createClient(ZKHelixManager.java:749)
	at org.apache.helix.manager.zk.ZKHelixManager.connect(ZKHelixManager.java:786)
	at org.apache.helix.integration.manager.MockParticipantManager.run(MockParticipantManager.java:98)
	at java.lang.Thread.run(Thread.java:745)

This exception is due to retrieving the cloud config would fail in local environment from 169.254.169.254 configured for cloud. This exception is a must for the test to succeed because if prevent ParticipantManager from creating liveInstance after autojoin().

If the run in GitHub testing machine running this unit test would be able to access `169.254.169.254`. This exception may not throw and thus the test may fail.

Search the log from failed test and we did not see this exception log. 
https://pipelines.actions.githubusercontent.com/sLVPVVo6suizPM5gIrHDigmaxapcQ7bGBvxdUFsAfGjAfNkqWG/_apis/pipelines/1/runs/188/signedlogcontent/3?urlExpires=2020-08-12T19%3A51%3A15.4635058Z&urlSigningMethod=HMACV1&urlSignature=5Z2u1uaZZHr3ScEm6sY7qj%2F77j%2FR4ArPftTUxejGaQg%3D

Thus, a strong support to the theory we need configure for sure to fail endpoint to retrieve cloud information.

```
   public void testAutoRegistrationShouldFailWhenWaitingResponse() throws Exception {
    // Create CloudConfig object and add to config
    CloudConfig.Builder cloudConfigBuilder = new CloudConfig.Builder();
    cloudConfigBuilder.setCloudEnabled(true);
    cloudConfigBuilder.setCloudProvider(CloudProvider.AZURE);
    cloudConfigBuilder.setCloudID("TestID");
    CloudConfig cloudConfig = cloudConfigBuilder.build();

    HelixManager manager = _participants[0];
    HelixDataAccessor accessor = manager.getHelixDataAccessor();

    _gSetupTool.addResourceToCluster(CLUSTER_NAME, db3, 60, "OnlineOffline",
        RebalanceMode.FULL_AUTO + "");
    _gSetupTool.rebalanceStorageCluster(CLUSTER_NAME, db3, 1);
    String instance3 = "localhost_279700";

    ConfigScope scope = new ConfigScopeBuilder().forCluster(CLUSTER_NAME).build();

    manager.getConfigAccessor().set(scope, ZKHelixManager.ALLOW_PARTICIPANT_AUTO_JOIN, "true");
    // Write the CloudConfig to Zookeeper
    PropertyKey.Builder keyBuilder = accessor.keyBuilder();
    accessor.setProperty(keyBuilder.cloudConfig(), cloudConfig);

    MockParticipantManager autoParticipant =
        new MockParticipantManager(ZK_ADDR, CLUSTER_NAME, instance3);
    autoParticipant.syncStart();

    Assert.assertTrue(null == manager.getHelixDataAccessor()                  -----------> failure
        .getProperty(accessor.keyBuilder().liveInstance(instance3)));
    try {
      manager.getConfigAccessor().getInstanceConfig(CLUSTER_NAME, instance3);
      fail(
          "Exception should be thrown because the instance cannot be added to the cluster due to the disconnection with Azure endpoint");
    } catch (HelixException e) {

    }
  }
```

- leak of one participant manager, minor

### LINKED ISSUE #1264: Change TestInstanceAutoJoin to adapt to cloud environment
**Describe the bug**
The previous test `testAutoRegistrationShouldFailWhenWaitingRespons` in TestInstanceAutoJoin will fail in github test, although it always pass locally. The reason is that the test was written with the assumption that it is only run in on-prem environment and auto registration will fail with no cloud response returned. However, github tests are run in cloud environment, and the auto registration will succeed. So we need to change the tests to make it adapt to both on-prem and cloud environment.

**To Reproduce**
Run the test in github.

**Expected behavior**
Auto registration succeeds and live instance is added.

### PR DESCRIPTION
**Pull request by zhangmeng916**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:

Fixed #1260  

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

The previous test testAutoRegistrationShouldFailWhenWaitingRespons in TestInstanceAutoJoin will fail in github test, although it always pass locally. The reason is that the test was written with the assumption that it is only run in on-prem environment and auto registration will fail with no cloud response returned. However, github tests are run in cloud environment, and the auto registration will succeed. This PR changed the test to make it adapt to both on-prem and cloud environment. 

### Tests

- [X] The following is the result of the "mvn test" command on the appropriate module:

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/test/java/org/apache/helix/integration/paticipant/TestInstanceAutoJoin.java [hunk 1]
```diff
@@ -110,14 +108,23 @@ public void testAutoRegistrationShouldFailWhenWaitingResponse() throws Exception
         new MockParticipantManager(ZK_ADDR, CLUSTER_NAME, instance3);
     autoParticipant.syncStart();
 
-    Assert.assertTrue(null == manager.getHelixDataAccessor()
-        .getProperty(accessor.keyBuilder().liveInstance(instance3)));
+    // if the test is run in cloud environment, auto registration will succeed and live instance
+    // will be added, otherwise, auto registration will fail and instance config will not be
+    // populated. An exception will be thrown.
     try {
       manager.getConfigAccessor().getInstanceConfig(CLUSTER_NAME, instance3);
```
Comment (kaisun2000) [inline helix-core/src/test/java/org/apache/helix/integration/paticipant/TestInstanceAutoJoin.java:115]: Why this line is needed?

**Comment by kaisun2000** [comment]
Let us see the results from github test run.

File: helix-core/src/test/java/org/apache/helix/integration/paticipant/TestInstanceAutoJoin.java (hunk 1 above)
Comment (zhangmeng916) [reply helix-core/src/test/java/org/apache/helix/integration/paticipant/TestInstanceAutoJoin.java:115]: This line is used to trigger the exception to be thrown as the instance config is not there.

**Comment by zhangmeng916** [comment]
> Let us see the results from github test run.

@kaisun2000 
Here's the result from github run:

[ERROR] Failures: 
[ERROR]   TestEnableCompression.testEnableCompressionResource:116 » ThreadTimeout Method...
[ERROR]   TestWagedRebalance.testChangeIdealState:303->validate:646 expected:<true> but was:<false>
[ERROR]   TestEnqueueJobs.testQueueJobsMaxCapacity » ThreadTimeout Method org.testng.int...
[INFO] 
[ERROR] Tests run: 1164, Failures: 3, Errors: 0, Skipped: 0

File: helix-core/src/test/java/org/apache/helix/integration/paticipant/TestInstanceAutoJoin.java (hunk 1 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/integration/paticipant/TestInstanceAutoJoin.java:115]: Got it. You are right.

**Comment by zhangmeng916** [comment]
This PR is ready to merge, approved by @kaisun2000 
Final commit message: 
Change TestInstanceAutoJoin to adapt to cloud environment

**Pull request merged**